anpr/license_plate_recognition.py

```diff
@@ -0,0 +1,337 @@
+
+        # Only return location on the image
+        return [location for _, location in largest_area_tuples]
+
+    def _pick_license_plate_location(
+        self, gray: np.array, candidates: list[ResultLocation]  # type: ignore
+    ) -> tuple[np.array, ResultLocation]:  # type: ignore
+        """
+        Pick the best estimation of the position of the license plate. Based on shape, contrast and presence of text.
+
+        @param gray: grayscale image to find license plate on.
+        @param candidates: list of locations with possible license plates.
+        @return: tuple with cropped image and ResultLocation of the candidate that best fits the properties of a license plate.
+        """
+
+        # initialize the license plate contour and ROI
+        lp_location = None
+        roi = None
+        # Variable to remember length of the longest string in license plate candidates
+        longest_text_length = 0
+
+        # loop over the license plate candidate contours
+        self._debug_print(f"Picking out of {len(candidates)} candidates.", 2)
+        for location in candidates:
+            # store the license plate contour and extract the
+            # license plate from the grayscale image and then
+            # threshold it
+            x, y = location.topLeft
+            h = location.height
+            w = location.width
+
+            licensePlate = gray[y : y + h, x : x + w]
+
+            roi_temp = cv2.threshold(
+                licensePlate, 0, 255, cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU
+            )[1]
+
+            # Select the license plate with the longest text on it
+            ocr_results = self.selection_ocr.get_text_from_image(roi_temp)  # Read text
+            self._debug_print(
+                f"Found Text on Candidate: {[r.text for r in ocr_results]}", 3
+            )
+
+            if len(ocr_results) > 0:
+                self._debug_imshow("License Plate Candidate", licensePlate, 3)
+                self._debug_imshow(
+                    "ROI",
+                    roi_temp,
+                    3,
+                    waitKey=True,
+                )
+                text_length = sum(len(r.text) for r in ocr_results)
+                if text_length is None or longest_text_length < text_length:
+                    longest_text_length = text_length
+                    lp_location = location
+                    roi = roi_temp
+
+        # return a 2-tuple of the license plate ROI and the contour
+        # associated with it
+        if roi is not None:
+            cv2.imshow("ROI", roi)
+        return roi, lp_location  # type: ignore
+
+    def find_and_ocr(
+        self, image_path: str, doSelection: bool = True
+    ) -> list[LicensePlateResult]:
+        """
+        Find license plate candidates and get the final text on it.
+
+        @param image: image to find license plate on
+        @param doSelection: boolean indicating whether the algorithm should first use the
```

Comment:
The doSelection parameter doesn't do anything anymore, so it should probably be removed to prevent confusion later. We can probably also just ignore the ANPR class and only use GoogleVisionOCR and the format matching seperatly as a lot of the code in this file won't be used.